README.md
smop-testsuite
==============

Large open-source matlab packages used as test data

TODO

   * YALMIP http://users.isy.liu.se/johanl/yalmip/pmwiki.php?n=Main.Download
   * SEDUMI http://sedumi.ie.lehigh.edu/


matlab2fmex/changelog
01/01/01 - Initial conception
06/11/01 - Initial release.
	 - made scalar convert option (want_ss=-1)
06/23/01 - fixed bug of elipses in disp statements.
07/02/01 - Fixed some bugs relating to bracket checking. Also implemented
	multi-output function. This does not support subscripts on the
	output variables, however.
07/02/01 - Released 0.2.
07/20/01 - Fixed number reading bug concerning exponents
08/14/01 - Added repmat, reshape, and linspace, fixed min bug
08/20/01 - Revised function substitution procedure to construct only those
	generic interfaces called for. Also fixed memory leak due to
	pointer function results by using temporary pointers in the main
	prg.
08/27/01 - Fixed the data type in fillc in matlabcallback to be type
	complex
08/27/01 - Added some safeguard for calling translated functions.
08/27/01 - Allowed converter parameters to be passed as function
	arguments.
08/29/01 - Changed tick mark feedback. Now executes faster.
08/29/01 - Removed brackets from 1x1 expressions.
08/30/01 - Updated mxs assignment criteria.
08/30/01 - Rewrote math operator decisions.
09/05/01 - Added multiple function capability.
09/07/01 - Released 0.5
09/11/01 - WTC terrorist attacks
09/13/01 - Can now have arbitrary colon lists in brackets 
	(e.g. [.1:.025:.9])
09/18/01 - Fixed bug converting while loops
09/19/01 - Added diag to converted functions
09/20/01 - Added svd, norm to converted functions
09/20/01 - Speeded up mxmult by using matmul directly
10/01/01 - Made improvements to scalar assignments.
10/01/01 - Fixed bug regarding detecting scalars
10/01/01 - Fixed problem converting multiple functions
10/01/01 - Added flag to disallow integers for local vars
10/08/01 - Released 0.51
10/18/01 - increased speed by changing updatefunstr
10/29/01 - Added a few functions -- bessels, airy, gamma, gammaln
11/07/01 - Fixed bug in making subscripts
11/07/01 - Can now call without explicitly specifying output var specs
11/07/01 - Eliminated [] flags requirement
11/07/01 - Fixed bug in multiple output mlcalls
11/09/01 - Added library call support for subroutines
11/09/01 - Fixed subroutines not reporting interfaces to main
11/11/01 - Made svd and eig double precision
11/13/01 - added detection of dos or unix os for system calls 
11/13/01 - Released 0.6
11/19/01 - Fixed bug with .^ translation
11/27/01 - Fixed bug involving CR on some Windows machines (RJHP)
12/10/01 - Added matlab2fmex_save to help with saving the workspace
01/21/02 - Fixed bugs dealing with multi-output matlab function calls
01/22/02 - Fixed log10 and log2 to handle complex numbers
01/28/02 - Fixed a bug which sized local variables incorrectly
01/28/02 - Corrected some incorrect type assignments
01/30/02 - Sped things up a bit by not traversing redundant loops
02/08/02 - Made colon expressions in subscripts more reliable
02/09/02 - Added Matlab's switch functionality
03/10/02 - Fixed dynamic output and local variable sizing
04/20/02 - Fixed a bug which required at least 1 local var
04/24/02 - Allowed complex variables to be function inputs
04/25/02 - Released 0.7
04/29/02 - Fixed a bug not requiring integers in some subscripts
07/11/02 - There has been ongoing development, though this file has not
        been updated. The major change has been in the interface. Explicit
	fortran pointers are no longer used, but Matlab library routines
	line mxCopyPtrToReal8 are used instead. This makes matlab2fmex
	available to a much broader range of compilers.
07/11/02 - Several bugs relating to var sizes in subroutines fixed.
07/11/02 - Released 0.8
07/20/02 - Added recursive subroutine/function capability
07/24/02 - Added support for fliplr, flipud, nnz
07/29/02 - Fixed a bug inserting extra commas for some [ ] assignments
07/30/02 - Added support for median and mean
08/03/02 - Improved trig interfaces (no longer uses slatec library)
08/03/02 - Much improvement for logical and other indexing of variables
08/04/02 - added better find functionality
08/05/02 - Released 0.85
09/24/02 - Again, ongoing development. Rewrote many core conversions to
        improve speed by utilizinf fortran intrinsics and minimizing
	allocatable arrays. Released 1.00.
01/29/04 - Fixed potential bug in detecting alpha machines
01/30/04 - Removed option for removing extra files after compiling
02/01/04 - Switched to f90 free format file type
02/01/04 - relocated helper mex modules to beginning of converted file
09/22/04 - modified so no longer have to pass array sizes
09/22/04 - mxCreateull -> mxCreateDoubleMatrix

ezyfit/efmenu.m
function efmenu(opt)
%EFMENU   Ezyfit menu
%   EFMENU adds or refreshes the Ezyfit menu for the current figure and
%   for all new figures.
%   EFMENU OFF removes the menu from all the figures and for all new
%   figures.
%
%   Careful: when the Ezyfit menu is active, all saved figure files (.FIG)
%   include the Ezyfit menu. This generates a problem when those figure
%   files are opened in a Matlab system which does not have the Ezyfit
%   toolbox installed. In order to remove the Ezyfit menu from the figure
%   file, use the function REMOVE_EFMENU_FIG.
%
%   If you want to always have the EzyFit menu in your figures, type
%   EFMENU INSTALL. This will create or update the 'startup.m' file in the
%   user directory of your Matlab installation. In addition, at
%   each Matlab restart, this will check the last version of the EzyFit
%   toolbox on the web (see CHECKUPDATE_EF).
%
%   In order to cancel the effect of EFMENU INSTALL, you need to edit the
%   'startup.m' file and to remove manually the 2 lines of code related to
%   the Ezyfit toolbox.
%
%   See also PLOTSAMPLE, SHOWFIT, EZFIT, UNDOFIT, RMFIT, EDITFIT,
%   LOADFIT, CHECKUPDATE_EF, REMOVE_EFMENU_FIG.

%   F. Moisy, moisy_at_fast.u-psud.fr
%   Revision: 1.80,  Date: 2012/07/02
%   This function is part of the EzyFit Toolbox

% History:
% 2005/12/06: v1.00, first version.
% 2005/12/16: v1.01, works with all new figures.
% 2006/01/12: v1.10, User defined fits improved.
% 2006/01/19: v1.20, User defined fits improved again.
% 2006/01/30: v1.21, bug fixed with selectfit. New fit fgauss.
% 2006/02/08: v1.30, use a new mat-file for user defined fits.
%                    polynomial orders as a submenu. New 'Edit Fit Coeff'.
%                    Check for update when launched from startup.m
% 2006/02/16: v1.31, item 'Show Fit Residual' added
% 2006/02/27: v1.40, option 'install' added.
% 2006/03/07: v1.41, option 'install' improved (append to startup.m), and
%                    bug fixed for check of the ml version
% 2006/04/10: v1.42, toobox renamed 'EzyFit'
% 2006/04/24: v1.43, bugs in option 'install' fixed (did not correctly
%                    recognize if the toolbox was already installed)
% 2006/09/06: v1.44, submenu for 'PlotSample'
% 2007/09/12: v1.45, new menu 'Refresh EzyFit menu'
% 2007/09/14: v1.46, bug fixed with 'install off' (case sensitive); options
%                    modified for getslope and showslope
% 2007/09/18: v1.50, new submenus for Gaussian fits
% 2008/03/06: v1.51, Ctrl-Z removed (conflicts)
% 2008/03/14: v1.60, selectfit removed for Ezyfit >= 2.30 (replaced by
%                    showfit in "Brushed" mode)
% 2009/02/04: v1.61, help text 'uninstall' added
% 2010/07/07: v1.70, help text 'remove_efmenu_fig' added
% 2012/06/28: v1.71, check update removed
% 2012/07/02: v1.80, change working directory, for Windows 7 compatibility
% 2012/08/01: v1.81, change for under MacOS X, Matlab 2012 (thanks J. Bagrow)



% check the matlab version:
% if str2double(version('-release'))<14,
%     error('EzyFit requires Matlab 7 (R14) or higher.');
% end

% if efmenu is called from the file startup.m, then calls
% the Check Update for the Ezyfit Toolbox:
st=dbstack;
if length(st)>=2
    if strcmp(st(2).name,'startup'),
        checkupdate_ef;
    end
end


if nargin==0
    opt='on';
end

% new v1.30:
userfit=loadfit('user');

switch lower(opt)
    case 'off'
        set(0,'DefaultFigureCreateFcn','');
        delete(findobj('Label','EzyFit')); % delete the menu from all the figures.
    case 'install'
        install_ef;
    case 'on'
        set(0,'DefaultFigureCreateFcn','efmenu');
        if get(0,'CurrentFigure')
            delete(findobj(gcf,'Label','EzyFit')); % delete the menu only from the current figure
            h=uimenu('Label','EzyFit');
            uimenu(h,'Label','Undo fit','Callback','undofit'); % Ctrl-Z removed march 7, 2008
            uimenu(h,'Label','Remove all fits','Callback','rmfit');

            %submenu showfit and selectfit:
            hm(1)=uimenu(h,'Separator','on','Label','Show Fit');
            menuname{1}='showfit';
            %hm(2)=uimenu(h,'Label','Select Fit');
            %menuname{2}='selectfit';
            for i=1:1, % only "showfit" is visible for EzyFit >=2.30
                uimenu(hm(i),'Label','Constant','Callback',['lastfit=' menuname{i} '(''cste'');']);
                hlin=uimenu(hm(i),'Label','Linear');
                uimenu(hlin,'Label','a*x','Callback',['lastfit=' menuname{i} '(''linear'');']);                
                uimenu(hlin,'Label','a*x+b','Callback',['lastfit=' menuname{i} '(''affine'');']);
                uimenu(hlin,'Label','a*(x-b)','Callback',['lastfit=' menuname{i} '(''affineshift'');']);
                hpol=uimenu(hm(i),'Label','Polynomial');
                for or=1:6,
                    uimenu(hpol,'Label',['Order ' num2str(or)],...
                        'Callback',['lastfit=' menuname{i} '(''poly' num2str(or) ''');']);
                end
                uimenu(hpol,'Label','Other...','Separator','on','Callback',['lastfit=' menuname{i} '(''poly'');']);
                hexp=uimenu(hm(i),'Label','Exponential');
                uimenu(hexp,'Label','a*exp(b*x)','Callback',['lastfit=' menuname{i} '(''exp'');']);
                uimenu(hexp,'Label','a*exp(b*x)+c','Callback',['lastfit=' menuname{i} '(''expc'');']);
                uimenu(hexp,'Label','a*exp(x/b)','Callback',['lastfit=' menuname{i} '(''expdiv'');']);
                uimenu(hexp,'Label','a*(1-exp(-x/b))','Callback',['lastfit=' menuname{i} '(''explim'');']);
                hlog=uimenu(hm(i),'Label','Logarithmic');
                uimenu(hlog,'Label','a*log(b*x)','Callback',['lastfit=' menuname{i} '(''log'');']);
                uimenu(hlog,'Label','a*log(x)+b','Callback',['lastfit=' menuname{i} '(''logc'');']);
                hpow=uimenu(hm(i),'Label','Power');
                uimenu(hpow,'Label','a*x^n','Callback',['lastfit=' menuname{i} '(''power'');']);
                uimenu(hpow,'Label','a*x^n+c','Callback',['lastfit=' menuname{i} '(''powerc'');']);
                uimenu(hpow,'Label','a*(x+b)^n','Callback',['lastfit=' menuname{i} '(''powershift'');']);
                hgauss=uimenu(hm(i),'Label','Gaussian');
                uimenu(hgauss,'Label','Centered Normalized Gaussian','Callback',['lastfit=' menuname{i} '(''cngauss'');']);
                uimenu(hgauss,'Label','Centered Non-normalized Gaussian','Callback',['lastfit=' menuname{i} '(''cfgauss'');']);
                uimenu(hgauss,'Label','Non-centered Normalized Gaussian','Callback',['lastfit=' menuname{i} '(''ngauss'');']);
                uimenu(hgauss,'Label','Non-centered Non-normalized Gaussian','Callback',['lastfit=' menuname{i} '(''fgauss'');']);
                hosc=uimenu(hm(i),'Label','Oscillations');
                uimenu(hosc,'Label','a*sin(b*x)','Callback',['lastfit=' menuname{i} '(''sin'');']);
                uimenu(hosc,'Label','a*cos(b*x)','Callback',['lastfit=' menuname{i} '(''cos'');']);
                uimenu(hosc,'Label','a*sin(b*x)+c','Callback',['lastfit=' menuname{i} '(''sinc'');']);
                uimenu(hosc,'Label','a*cos(b*x)+c','Callback',['lastfit=' menuname{i} '(''cosc'');']);
                uimenu(hosc,'Label','a*sin(b*x+phi)','Callback',['lastfit=' menuname{i} '(''sinphi'');']);
                uimenu(hosc,'Label','a*sin(b*x+phi)+c','Callback',['lastfit=' menuname{i} '(''sinphic'');']);
                uimenu(hm(i),'Label','Nearest','Separator','on','Callback',[menuname{i} '(''nearest'');']);
                uimenu(hm(i),'Label','Spline','Callback',[menuname{i} '(''spline'');']);
                uimenu(hm(i),'Label','Cubic','Callback',[menuname{i} '(''cubic'');']);
                
                for j=1:length(userfit)
                    if j==1,
                        uimenu(hm(i),'Label',['#' num2str(j) ': ' userfit(j).name],...
                            'Separator','on','Callback',['lastfit=' menuname{i} '(''' userfit(j).eq ''');']);
                    else
                        uimenu(hm(i),'Label',['#' num2str(j) ': ' userfit(j).name],...
                            'Callback',['lastfit=' menuname{i} '(''' userfit(j).eq ''');']);
                    end
                end
                uimenu(hm(i),'Label','Other...','Separator','on','Callback',[menuname{i} ';']);
            end

            %submenu edit fit:
            hef=uimenu(h,'Label','Edit User Fit');
            for j=1:length(userfit)
                uimenu(hef,'Label',['#' num2str(j) ': ' userfit(j).name],'Callback',['editfit(' num2str(j) ')']);
            end
            uimenu(hef,'Label','New User Fit...','Separator','on','Callback','editfit');
            uimenu(hef,'Label','Reset','Separator','on','Callback',...
                'if strcmp(questdlg(''Are you sure you want to reset the user defined fits?'',''Reset User Fits'',''No''),''Yes''), delete(''userfit.mat''); efmenu; end');

            
            % remainder of the menu:
            uimenu(h,'Label','Refresh EzyFit menu','Callback','efmenu');
            uimenu(h,'Label','Edit Fit Coefficients','Callback','editcoeff');
            uimenu(h,'Label','Show Fit Residuals','Callback','showresidual');

            uimenu(h,'Label','Get Slope','Separator','on','Callback','getslope(''equation'',''dialog'')','Accelerator','G');
            uimenu(h,'Label','Show Slope...','Callback','showslope([],''dialog'')');

            hps=uimenu(h,'Label','Plot Sample','Separator','on');            
            uimenu(hps,'Label','Linear','Callback','plotsample linear');
            uimenu(hps,'Label','Constant','Callback','plotsample cste');
            uimenu(hps,'Label','3 polynomial curves','Callback','plotsample poly2');
            uimenu(hps,'Label','Oscillations','Callback','plotsample osc');            
            uimenu(hps,'Label','Damped oscillations','Callback','plotsample damposc');            
            uimenu(hps,'Label','Exponential decay','Callback','plotsample exp');
            uimenu(hps,'Label','Histogram with one peak','Callback','plotsample hist');
            uimenu(hps,'Label','Histogram with two peaks','Callback','plotsample hist2');
            uimenu(hps,'Label','Power law','Callback','plotsample power');
            uimenu(hps,'Label','Power law with a cutoff','Callback','plotsample powco');          
            
            hqc=uimenu(h,'Label','Axis','Separator','on');
            uimenu(hqc,'Label','Swap X-axis lin <-> log','Callback','swx');
            uimenu(hqc,'Label','Swap Y-axis lin <-> log','Callback','swy');
            uimenu(hqc,'Label','Swap both X- and Y- axis','Callback','sw','Accelerator','B');
            uimenu(hqc,'Label','Include the origin','Callback','axis0');
            uimenu(hqc,'Label','Center the origin','Callback','axisc');
            uimenu(hqc,'Label','Enlarge to nearest power of 10','Callback','axisl');
            
            uimenu(h,'Label','Edit Default Settings...','Separator','on','Callback','edit fitparam');
            uimenu(h,'Label','Help EzyFit','Callback','docezyfit');
            uimenu(h,'Label','Check for update','Callback','checkupdate_ef(''dialog'')');
            uimenu(h,'Label','EzyFit Home Page','Callback','web http://www.fast.u-psud.fr/ezyfit -browser');
            uimenu(h,'Label','About EzyFit','Callback','about_ef(''dialog'')');
        end
    otherwise
        error('Unknown option');
end


% -----------------------------------------
% Subfunction 'efmenu install'


function install_ef

% This location no longer works for Windows 7:
%sufile = fullfile(matlabroot,'toolbox','local','startup.m'); 

sufolder = strrep(userpath,';','');
sufolder = strrep(sufolder,':',''); % new v1.81, thanks J. Bagrow
sufile = fullfile(sufolder,'startup.m');

ft = [];
if exist(sufile,'file')
    try
        ft = textread(sufile,'%s');
    catch
        ft = [];
    end
end
if ~isempty(ft)
    for num = 1:length(ft)
        if strfind(ft{num},'efmenu'),
            disp('The EzyFit menu is already installed in your startup.m file:');
            disp(sufile);
            return
        end
    end
    %copyfile(sufile,fullfile(matlabroot,'toolbox','local','startup_previous.m'));
    copyfile(sufile,fullfile(sufolder,'startup_previous.m'));
    fid=fopen(sufile,'a');
    fprintf(fid,'%s\n',' ');
    fprintf(fid,'%s\n',['%   These lines have been added by ''efmenu install'' (' datestr(now) '):']);
    fprintf(fid,'%s\n','efmenu;   % Includes the EzyFit menu for all new figure.');
    fprintf(fid,'%s\n\n','fprintf('' To get started with the EzyFit toolbox, select <a href="matlab:docezyfit">EzyFit</a> from the Help browser.\n\n'');');    
    fclose(fid);
    
    disp('The EzyFit menu has been correctly installed.');
    disp('To get started, select <a href="matlab:docezyfit">EzyFit</a> from the Help browser.');

else % creates a new startup.m file and re-calls install_ef:
    fid=fopen(sufile,'w');
    fprintf(fid,'%s\n','%STARTUP   Startup file');
    fprintf(fid,'%s\n','%   This file is executed when MATLAB starts up.');
    fprintf(fid,'%s\n',' ');
    fclose(fid);
    install_ef;
end
